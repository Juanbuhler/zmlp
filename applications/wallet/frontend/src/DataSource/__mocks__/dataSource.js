const dataSource = {
  id: '2f0de857-95fd-120e-85f3-0242ac120002',
  projectId: '00000000-0000-0000-0000-000000000000',
  name: 'test',
  uri: 'gs://zorroa-dev-data/images',
  fileTypes: ['jpg'],
  credentials: '',
  modules: ['1810b623-4b95-11ea-8b23-0242ac120004'],
  timeCreated: 1581298667707,
  timeModified: 1581298667707,
  actorCreated: 'admin-key',
  actorModified: 'admin-key',
  url:
    'http://localhost/api/v1/projects/00000000-0000-0000-0000-000000000000/data_sources/2f0de857-95fd-120e-85f3-0242ac120002/',
}

export default dataSource
