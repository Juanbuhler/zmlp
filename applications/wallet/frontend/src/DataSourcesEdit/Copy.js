import { constants, colors } from '../Styles'

const DataSourcesEditCopy = () => {
  return (
    <p
      css={{
        marginTop: 0,
        maxWidth: constants.paragraph.maxWidth,
        color: colors.structure.zinc,
      }}
    >
      Additional file types or analysis modules can be added to this data
      source. All previous selections will remain constant and cannot be
      modified.
    </p>
  )
}

export default DataSourcesEditCopy
