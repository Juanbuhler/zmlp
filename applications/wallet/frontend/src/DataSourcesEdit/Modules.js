import PropTypes from 'prop-types'

import checkboxOptionShape from '../Checkbox/optionShape'

import { colors, spacing, constants } from '../Styles'

import CheckmarkSvg from '../Icons/checkmark.svg'

import Accordion, { CHECKMARK_WIDTH } from '../Accordion'
import CheckboxTable from '../Checkbox/Table'

const DataSourcesEditModules = ({
  module: { provider, description, categories },
  onClick,
}) => {
  return (
    <Accordion
      title={
        <>
          <CheckmarkSvg
            width={CHECKMARK_WIDTH}
            css={{ color: colors.key.one, marginRight: spacing.normal }}
          />
          {provider}
        </>
      }>
      <>
        <p
          css={{
            color: colors.structure.zinc,
            margin: 0,
            paddingTop: spacing.base,
            paddingBottom: spacing.normal,
            maxWidth: constants.paragraph.maxWidth,
          }}>
          {description}
        </p>
        {categories.map(category => (
          <CheckboxTable
            key={category.name}
            category={category}
            onClick={onClick}
          />
        ))}
      </>
    </Accordion>
  )
}

DataSourcesEditModules.propTypes = {
  module: PropTypes.shape({
    provider: PropTypes.string.isRequired,
    logo: PropTypes.node.isRequired,
    description: PropTypes.node.isRequired,
    categories: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        options: PropTypes.arrayOf(PropTypes.shape(checkboxOptionShape))
          .isRequired,
      }).isRequired,
    ).isRequired,
  }).isRequired,
  onClick: PropTypes.func.isRequired,
}

export default DataSourcesEditModules
