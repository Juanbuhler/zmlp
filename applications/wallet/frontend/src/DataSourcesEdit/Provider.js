import PropTypes from 'prop-types'

import { colors, spacing, constants } from '../Styles'

import Accordion from '../Accordion'
import CheckboxTable from '../Checkbox/Table'

const IMG_HEIGHT = 32

const DataSourcesEditProvider = ({
  provider: { name, logo, description, categories },
  modules,
  onClick,
}) => {
  return (
    <Accordion title={<img src={logo} alt={name} height={IMG_HEIGHT} />}>
      <>
        <p
          css={{
            color: colors.structure.zinc,
            margin: 0,
            paddingTop: spacing.base,
            paddingBottom: spacing.normal,
            maxWidth: constants.paragraph.maxWidth,
          }}
        >
          {description}
        </p>
        {categories.map((category) => (
          <CheckboxTable
            key={category.name}
            category={{
              name: category.name,
              options: category.modules.map((module) => {
                return {
                  value: module.name,
                  label: module.description,
                  initialValue: modules[module.name] || false,
                  isDisabled: module.restricted,
                }
              }),
            }}
            onClick={onClick}
          />
        ))}
      </>
    </Accordion>
  )
}

DataSourcesEditProvider.propTypes = {
  provider: PropTypes.shape({
    name: PropTypes.string.isRequired,
    logo: PropTypes.node.isRequired,
    description: PropTypes.node.isRequired,
    categories: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        modules: PropTypes.arrayOf(PropTypes.shape()).isRequired,
      }).isRequired,
    ).isRequired,
  }).isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  modules: PropTypes.object.isRequired,
  onClick: PropTypes.func.isRequired,
}

export default DataSourcesEditProvider
