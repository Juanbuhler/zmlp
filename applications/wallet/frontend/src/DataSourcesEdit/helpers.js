import Router from 'next/router'

import {
  fetcher,
  revalidate,
  getQueryString,
  parseResponse,
} from '../Fetch/helpers'

export const getInitialModules = ({
  initialState: { modules: existingModules },
  providers,
}) => {
  const results = {}

  providers.forEach(({ categories }) =>
    categories.forEach(({ modules }) =>
      modules.forEach(({ id, name }) => {
        if (existingModules.includes(id)) {
          results[name] = true
        }
      }),
    ),
  )

  return results
}

export const onSubmit = async ({
  dispatch,
  projectId,
  dataSourceId,
  state: { name, uri, fileTypes, modules, credentials },
}) => {
  dispatch({ isLoading: true, errors: {} })

  try {
    const { jobId } = await fetcher(
      `/api/v1/projects/${projectId}/data_sources/${dataSourceId}/`,
      {
        method: 'PUT',
        body: JSON.stringify({
          name,
          uri,
          fileTypes: Object.keys(fileTypes).filter((f) => fileTypes[f]),
          modules: Object.keys(modules).filter((m) => modules[m]),
          credentials,
        }),
      },
    )

    revalidate({
      key: `/api/v1/projects/${projectId}/data_sources/${dataSourceId}/`,
      paginated: false,
    })

    await revalidate({
      key: `/api/v1/projects/${projectId}/data_sources/`,
      paginated: true,
    })

    const queryString = getQueryString({
      action: 'edit-datasource-success',
      jobId,
    })

    Router.push(
      `/[projectId]/data-sources${queryString}`,
      `/${projectId}/data-sources`,
    )
  } catch (response) {
    const errors = await parseResponse({ response })

    dispatch({ isLoading: false, errors })
  }
}
