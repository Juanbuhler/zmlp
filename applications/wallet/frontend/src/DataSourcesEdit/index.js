import Head from 'next/head'

import PageTitle from '../PageTitle'
import Tabs from '../Tabs'
import SuspenseBoundary, { ROLES } from '../SuspenseBoundary'

import DataSourcesEditContent from './Content'

const DataSourcesEdit = () => {
  return (
    <>
      <Head>
        <title>Edit Data Source</title>
      </Head>

      <PageTitle>Data Sources</PageTitle>

      <Tabs
        tabs={[
          { title: 'View all', href: '/[projectId]/data-sources' },
          {
            title: 'Create Data Source',
            href: '/[projectId]/data-sources/add',
          },
          {
            title: 'Edit Data Source',
            href: '/[projectId]/data-sources/[dataSourceId]/edit',
          },
        ]}
      />

      <SuspenseBoundary role={ROLES.ML_Tools}>
        <DataSourcesEditContent />
      </SuspenseBoundary>
    </>
  )
}

export default DataSourcesEdit
