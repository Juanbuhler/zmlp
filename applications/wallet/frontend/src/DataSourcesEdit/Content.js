import useSWR from 'swr'
import { useRouter } from 'next/router'

import { FILE_TYPES } from '../DataSourcesAdd/helpers'

import DataSourcesForm from '../DataSourcesForm'

const DataSourcesEditContent = () => {
  const {
    query: { projectId, dataSourceId },
  } = useRouter()

  const {
    data: { name, uri, credential, fileTypes, modules },
  } = useSWR(`/api/v1/projects/${projectId}/data_sources/${dataSourceId}`)

  const groupedFileTypes = FILE_TYPES.reduce(
    (accumulator, { value, identifier }) => {
      const matchFileType = fileTypes.includes(identifier)

      if (matchFileType) {
        accumulator[value] = true
      }

      return accumulator
    },
    {},
  )

  const initialState = {
    name,
    uri,
    credential,
    fileTypes: groupedFileTypes,
    modules,
    errors: { global: '' },
  }

  return <DataSourcesForm initialState={initialState} />
}

export default DataSourcesEditContent
