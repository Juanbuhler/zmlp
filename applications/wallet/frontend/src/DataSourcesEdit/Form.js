import { useReducer } from 'react'
import { useRouter } from 'next/router'
import useSWR from 'swr'
import Link from 'next/link'
import PropTypes from 'prop-types'

import { constants, spacing, typography } from '../Styles'

import Form from '../Form'
import SectionTitle from '../SectionTitle'
import SectionSubTitle from '../SectionSubTitle'
import Input, { VARIANTS as INPUT_VARIANTS } from '../Input'
import Button, { VARIANTS as BUTTON_VARIANTS } from '../Button'
import FlashMessage, { VARIANTS as FLASH_VARIANTS } from '../FlashMessage'
import { VARIANTS as CHECKBOX_VARIANTS } from '../Checkbox'
import ButtonGroup from '../Button/Group'
import CheckboxGroup from '../Checkbox/Group'
import Toggletip from '../Toggletip'

import { FILE_TYPES } from '../DataSourcesAdd/helpers'

import DataSourcesAddAutomaticAnalysis from '../DataSourcesAdd/AutomaticAnalysis'

import DataSourcesEditProvider from './Provider'
import DataSourcesEditCopy from './Copy'

import { getInitialModules, onSubmit } from './helpers'

const reducer = (state, action) => ({ ...state, ...action })

const DataSourcesEditForm = ({ initialState }) => {
  const {
    query: { projectId, dataSourceId },
  } = useRouter()

  const {
    data: { results: providers },
  } = useSWR(`/api/v1/projects/${projectId}/providers/`)

  const initialModules = getInitialModules({
    initialState,
    providers,
  })

  const [state, dispatch] = useReducer(reducer, {
    ...initialState,
    modules: initialModules,
  })

  const { errors, fileTypes, name, uri } = state

  return (
    <>
      {errors.global && (
        <div
          css={{
            display: 'flex',
            paddingTop: spacing.base,
            marginBottom: -spacing.base,
          }}
        >
          <FlashMessage variant={FLASH_VARIANTS.ERROR}>
            {errors.global}
          </FlashMessage>
        </div>
      )}

      <Form style={{ width: 'auto' }}>
        <DataSourcesEditCopy />

        <div
          css={{
            width: constants.form.maxWidth,
            paddingBottom: spacing.comfy,
          }}
        >
          <SectionTitle>Data Source Name </SectionTitle>

          <Input
            autoFocus
            id="name"
            variant={INPUT_VARIANTS.SECONDARY}
            label="Name"
            type="text"
            value={name}
            onChange={({ target: { value } }) =>
              dispatch({
                name: value,
              })
            }
            hasError={!!errors.name || !name}
            errorMessage={errors.name || (!name ? 'Name cannot be empty' : '')}
          />

          <SectionTitle>{`Storage Address: ${uri}`}</SectionTitle>
        </div>

        <CheckboxGroup
          legend={
            <>
              Add Additional File Types
              <Toggletip openToThe="right" label="Supported File Types">
                <div
                  css={{
                    fontSize: typography.size.regular,
                    lineHeight: typography.height.regular,
                  }}
                >
                  <h3
                    css={{
                      fontSize: typography.size.regular,
                      lineHeight: typography.height.regular,
                      paddingBottom: spacing.base,
                    }}
                  >
                    Supported File Types
                  </h3>
                  {FILE_TYPES.map(({ value, extensions }) => (
                    <div key={value} css={{ paddingBottom: spacing.base }}>
                      <h4>{value}:</h4>
                      {extensions}
                    </div>
                  ))}
                </div>
              </Toggletip>
            </>
          }
          description={
            <div>
              Additional file types can be added to this data source. Previous
              selections cannot be removed.
            </div>
          }
          onClick={(fileType) =>
            dispatch({ fileTypes: { ...fileTypes, ...fileType } })
          }
          options={FILE_TYPES.map(({ value, label, legend, icon }) => ({
            value,
            label,
            icon: <img src={icon} alt={label} width="40px" />,
            legend,
            initialValue: !!fileTypes[value],
            isDisabled: !!initialState.fileTypes[value],
          }))}
          variant={CHECKBOX_VARIANTS.SECONDARY}
        />

        <div css={{ height: spacing.base }} />

        <SectionTitle>Add Additional Analysis</SectionTitle>

        <SectionSubTitle>
          Additional analysis can be added to this data source. Previous
          selections cannot be removed.
        </SectionSubTitle>

        <DataSourcesAddAutomaticAnalysis />

        {providers.map((provider) => (
          <DataSourcesEditProvider
            key={provider.name}
            provider={provider}
            initialModules={initialModules}
            modules={state.modules}
            onClick={(module) =>
              dispatch({ modules: { ...state.modules, ...module } })
            }
          />
        ))}

        <ButtonGroup>
          <Link
            href="/[projectId]/data-sources"
            as={`/${projectId}/data-sources`}
            passHref
          >
            <Button variant={BUTTON_VARIANTS.SECONDARY}>Cancel</Button>
          </Link>
          <Button
            type="submit"
            variant={BUTTON_VARIANTS.PRIMARY}
            onClick={() =>
              onSubmit({ dispatch, projectId, dataSourceId, state })
            }
            isDisabled={!name || state.isLoading}
          >
            {state.isLoading ? 'Updating...' : 'Update Data Source'}
          </Button>
        </ButtonGroup>
      </Form>
    </>
  )
}

DataSourcesEditForm.propTypes = {
  initialState: PropTypes.shape({
    name: PropTypes.string.isRequired,
    uri: PropTypes.string.isRequired,
    fileTypes: PropTypes.shape({
      Images: PropTypes.bool,
      Videos: PropTypes.bool,
      Documents: PropTypes.bool,
    }).isRequired,
    modules: PropTypes.arrayOf(PropTypes.string).isRequired,
    errors: PropTypes.shape({ global: PropTypes.string }).isRequired,
  }).isRequired,
}

export default DataSourcesEditForm
