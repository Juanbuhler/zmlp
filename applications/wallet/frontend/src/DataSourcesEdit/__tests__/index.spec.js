import TestRenderer from 'react-test-renderer'

import mockUser from '../../User/__mocks__/user'

import dataSource from '../../DataSource/__mocks__/dataSource'

import User from '../../User'

import DataSourcesEdit from '..'

jest.mock('../Form', () => 'DataSourcesEditForm')

const PROJECT_ID = '76917058-b147-4556-987a-0a0f11e46d9b'
const DATA_SOURCE_ID = dataSource.id

describe('<DataSourcesEdit />', () => {
  it('should render properly', () => {
    require('next/router').__setUseRouter({
      pathname: '/[projectId]/data-sources/add',
      query: { projectId: PROJECT_ID, dataSourceId: DATA_SOURCE_ID },
    })

    require('swr').__setMockUseSWRResponse({ data: dataSource })

    const component = TestRenderer.create(
      <User initialUser={mockUser}>
        <DataSourcesEdit />
      </User>,
    )

    expect(component.toJSON()).toMatchSnapshot()
  })
})
