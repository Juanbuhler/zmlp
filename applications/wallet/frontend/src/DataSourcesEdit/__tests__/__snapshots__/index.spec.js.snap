// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataSourcesEdit /> should render properly 1`] = `
Array [
  .emotion-0 {
  font-size: 18px;
  line-height: 21px;
  font-weight: 400;
  padding-top: 24px;
  padding-bottom: 16px;
}

<h2
    className="emotion-0"
  >
    Data Sources
  </h2>,
  .emotion-7 {
  padding: 16px;
  padding-left: 0;
  padding-right: 0;
}

.emotion-6 {
  list-style-type: none;
  padding: 0;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 1px solid #4a4a4a;
}

.emotion-1 {
  padding-right: 16px;
  margin-bottom: -1px;
}

.emotion-0 {
  border: 0 #73b61c solid;
  border-bottom-width: 0;
  color: #b3b3b3;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-top: 8px;
  padding-bottom: 8px;
  text-transform: uppercase;
  font-size: 17px;
  line-height: 20px;
  font-weight: 700;
}

.emotion-0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #73b61c;
}

.emotion-4 {
  border: 0 #73b61c solid;
  border-bottom-width: 2px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-top: 8px;
  padding-bottom: 8px;
  text-transform: uppercase;
  font-size: 17px;
  line-height: 20px;
  font-weight: 700;
}

.emotion-4:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<nav
    className="emotion-7"
  >
    <ul
      className="emotion-6"
    >
      <li
        className="emotion-1"
      >
        <a
          className="emotion-0"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/data-sources"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          View all
        </a>
      </li>
      <li
        className="emotion-1"
      >
        <a
          className="emotion-0"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/data-sources/add"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Add Data Source
        </a>
      </li>
      <li
        className="emotion-1"
      >
        <a
          className="emotion-4"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/data-sources/2f0de857-95fd-120e-85f3-0242ac120002/edit"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Edit Data Source
        </a>
      </li>
    </ul>
  </nav>,
  <DataSourcesForm
    initialState={
      Object {
        "credential": undefined,
        "errors": Object {
          "global": "",
        },
        "fileTypes": Object {
          "images": true,
        },
        "modules": Array [
          "1810b623-4b95-11ea-8b23-0242ac120004",
        ],
        "name": "test",
        "uri": "gs://zorroa-dev-data/images",
      }
    }
  />,
]
`;
